anvil/src/focus.rs

```diff
@@ -14,46 +12,80 @@ pub use smithay::{
     utils::{IsAlive, Serial},
     wayland::seat::WaylandFocus,
 };
+use smithay::{
+    desktop::Window,
+    input::{
+        pointer::{
+            GestureHoldBeginEvent, GestureHoldEndEvent, GesturePinchBeginEvent, GesturePinchEndEvent,
+            GesturePinchUpdateEvent, GestureSwipeBeginEvent, GestureSwipeEndEvent, GestureSwipeUpdateEvent,
+        },
+        touch::TouchTarget,
+    },
+};
 
 use crate::{
-    shell::WindowElement,
+    shell::{WindowElement, SSD},
     state::{AnvilState, Backend},
 };
 
 #[derive(Debug, Clone, PartialEq)]
-pub enum FocusTarget {
-    Window(WindowElement),
+pub enum KeyboardFocusTarget {
+    Window(Window),
+    #[cfg(feature = "xwayland")]
+    X11Surface(X11Surface),
     LayerSurface(LayerSurface),
     Popup(PopupKind),
 }
 
-impl IsAlive for FocusTarget {
+impl IsAlive for KeyboardFocusTarget {
     fn alive(&self) -> bool {
         match self {
-            FocusTarget::Window(w) => w.alive(),
-            FocusTarget::LayerSurface(l) => l.alive(),
-            FocusTarget::Popup(p) => p.alive(),
+            KeyboardFocusTarget::Window(w) => w.alive(),
+            #[cfg(feature = "xwayland")]
+            KeyboardFocusTarget::X11Surface(s) => s.alive(),
+            KeyboardFocusTarget::LayerSurface(l) => l.alive(),
+            KeyboardFocusTarget::Popup(p) => p.alive(),
         }
     }
 }
 
-impl From<FocusTarget> for WlSurface {
-    fn from(target: FocusTarget) -> Self {
+#[derive(Debug, Clone, PartialEq)]
+pub enum PointerFocusTarget {
+    WlSurface(WlSurface),
+    #[cfg(feature = "xwayland")]
+    X11Surface(X11Surface),
+    SSD(SSD),
+}
+
+impl IsAlive for PointerFocusTarget {
+    fn alive(&self) -> bool {
+        match self {
+            PointerFocusTarget::WlSurface(w) => w.alive(),
+            #[cfg(feature = "xwayland")]
+            PointerFocusTarget::X11Surface(w) => w.alive(),
+            PointerFocusTarget::SSD(x) => x.alive(),
+        }
+    }
+}
+
+impl From<PointerFocusTarget> for WlSurface {
+    fn from(target: PointerFocusTarget) -> Self {
         target.wl_surface().unwrap()
     }
 }
 
-impl<BackendData: Backend> PointerTarget<AnvilState<BackendData>> for FocusTarget {
+impl<BackendData: Backend> PointerTarget<AnvilState<BackendData>> for PointerFocusTarget {
     fn enter(
         &self,
         seat: &Seat<AnvilState<BackendData>>,
         data: &mut AnvilState<BackendData>,
         event: &MotionEvent,
     ) {
         match self {
-            FocusTarget::Window(w) => PointerTarget::enter(w, seat, data, event),
-            FocusTarget::LayerSurface(l) => PointerTarget::enter(l, seat, data, event),
-            FocusTarget::Popup(p) => PointerTarget::enter(p.wl_surface(), seat, data, event),
+            PointerFocusTarget::WlSurface(w) => PointerTarget::enter(w, seat, data, event),
+            #[cfg(feature = "xwayland")]
+            PointerFocusTarget::X11Surface(w) => PointerTarget::enter(w, seat, data, event),
```

Comment:
If `Window`/`LayerSurface`/`Popup` can be combined as just `WlSurface`, shouldn't that also cover `X11Surface`?

It seems `PointerTarget<D> for X11Surface` just calls `PointerTarget` methods on the underlying `WlSurface`, if it exists.

Reply:
Yes, that would be possible, but it felt strange to leave the `KeyboardTarget` impl there (which does a few x related stuff) and only remove the `PointerTarget` impl.

Reply:
Ah, fair point.